browser/ui/webui/settings/brave_import_bulk_data_handler.cc

```diff
@@ -0,0 +1,174 @@
+  // has been loaded, or with a nullptr otherwise.
+  profile_manager->LoadProfileByPath(
+      profile_path, false,
+      base::BindOnce(
+          [](ProfileReadyCallback profile_callback, const std::u16string& name,
+             Profile* existing_profile) {
+            // Existing profile loaded, reuse it.
+            if (existing_profile) {
+              std::move(profile_callback).Run(existing_profile);
+              return;
+            }
+            // Asynchronously creates a new profile in the next available
+            // multiprofile directory. Directories are named "profile_1",
+            // "profile_2", etc., in sequence of creation.
+            auto avatar_index =
+                base::RandInt(profiles::GetModernAvatarIconStartIndex(),
+                              profiles::GetDefaultAvatarIconCount());
+            ProfileManager::CreateMultiProfileAsync(
+                name, avatar_index, false,
+                base::BindOnce(
+                    [](ProfileReadyCallback initialized_callback,
+                       Profile* created_profile) {
+                      CHECK(created_profile);
+                      // Migrate welcome page flag to new profiles.
+                      created_profile->GetPrefs()->SetBoolean(
+                          prefs::kHasSeenWelcomePage, true);
+                      std::move(initialized_callback).Run(created_profile);
+                    },
+                    std::move(profile_callback)));
+          },
+          std::move(profile_ready_callback), name));
+}
+
+void BraveImportBulkDataHandler::HandleImportDataBulk(
+    const base::Value::List& args) {
+  CHECK_GE(args.size(), 2u);
+  const auto& list = args[0].GetList();
+  // Bulk profiles import assumes new profiles will be created on our side if
+  // they do not exist.
+  const base::Value& types = args[1];
+  for (const auto& it : list) {
+    int browser_index = it.GetInt();
+    importing_profiles_.insert(browser_index);
+    base::Value::List single_profile_args;
+    single_profile_args.Append(base::Value(browser_index));
+    single_profile_args.Append(types.Clone());
+    BraveImportDataHandler::HandleImportData(single_profile_args);
+  }
+}
+
+absl::optional<int> BraveImportBulkDataHandler::GetProfileIndex(
+    const importer::SourceProfile& source_profile) {
+  for (auto index : importing_profiles_) {
+    const auto& profile = GetSourceProfileAt(index);
+    if (profile.source_path == source_profile.source_path) {
+      return index;
+    }
+  }
+  return absl::nullopt;
+}
+
+void BraveImportBulkDataHandler::StartImport(
+    const importer::SourceProfile& source_profile,
+    uint16_t imported_items) {
+  // If profile is not from the bulk import request fallback to single profile
+  // import.
+  if (!GetProfileIndex(source_profile).has_value()) {
+    BraveImportDataHandler::StartImport(source_profile, imported_items);
+    return;
+  }
+  if (!imported_items)
```

Comment:
I think this could be located at the start of this method?

Reply:
done